docs/spec.md

```diff
@@ -2117,9 +2117,11 @@ tensor and produces a `result` tensor.
 #### Semantics
 
 Encapsulates an implementation-defined operation `call_target_name` that takes
-`inputs` and `called_computations` and produces `results`. `has_side_effect`,
-`backend_config` and `api_version` may be used to provide additional
-implementation-defined metadata.
+`inputs` and `called_computations` and produces `results`. The metadata
+`has_side_effect` is used to prevent optimizations like speculation (e.g
+loop-invariant code motion) and dead-code elimination. `backend_config`
```

Comment:
I'm not sure I understand from this description what program transformations are blocked by `has_side_effects`.

For comparison, [the spec of OptimizationBarrierOp](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#optimization_barrier) is very clear - program transformations cannot move operations across the barrier. So, all transformations that do this are illegal, and all transformations that don't do this are legal, as far as this operation is concerned.

Can we have a similarly clear criterion here?